cloud-engineering-with-gcp/python/content/lab-02/README.md

```diff
@@ -0,0 +1,199 @@
+
+Run `pulumi up` in your project directory. You should see something like this:
+
+```bash
+$ pulumi up
+Previewing update (dev)
+
+View Live: https://app.pulumi.com/jaxxstorm/static_site/dev/previews/b5ce861f-16ac-4930-ada4-1e9547e02eb1
+
+     Type                                       Name             Plan
+ +   pulumi:pulumi:Stack                        static_site-dev  create
+ +   ├─ gcp:storage:Bucket                      website          create
+ +   ├─ gcp:storage:DefaultObjectAccessControl  website          create
+ +   ├─ gcp:storage:BucketObject                404.html         create
+ +   └─ gcp:storage:BucketObject                index.html       create
+
+Resources:
+    + 5 to create
+
+Do you want to perform this update?  [Use arrows to move, enter to select, type to filter]
+  yes
+> no
+```
+
+You can examine the details of the resources that will be created. When you're happy, move the arrow to `yes` and watch as Pulumi creates your resources!
+
+## Step 5 &mdash; Export the Bucket URL
+
+Our final step is to build our static site URL and add it as a [stack output](https://www.pulumi.com/learn/building-with-pulumi/stack-outputs/).
+
+First, we assemble the output's value using `pulumi.Output.concat`:
+
+```python
+static_site_url = pulumi.Output.concat(
+    "https://storage.googleapis.com/", bucket.name, "/index.html")
+```
+
+We use `pulumi.Output.concat` instead of standard Python string concatenation because `bucket.name` is a Pulumi output - a value that isn't known until after a resource has been created. For more information on Pulumi inputs and outputs, see [Inputs and Outputs](https://www.pulumi.com/docs/intro/concepts/inputs-outputs/) in the Pulumi docs.
+
+Now we can export the value as a stack export, which will allow us to view its value after our Pulumi program runs and our infrastructure has been provisioned:
+
+```python
+pulumi.export("static_site_url", static_site_url)
+```
+
+We can obtain the value of our URL by running `pulumi up` again:
+
+```bash
+pulumi up
+```
+
+And we can now view our website:
+
+```bash
+curl $(pulumi stack output static_site_url)
+```
+
+At the end of this lab, your `__main__.py` should look like this:
+
+```python
+"""A Python Pulumi program"""
+
+import os
+
+import pulumi
+import pulumi_gcp as gcp
+
+bucket = gcp.storage.Bucket(
+    "website",
+    location="US"
+)
+
+acl = gcp.storage.DefaultObjectAccessControl(
+    'website',
+    bucket=bucket.name,
+    role="READER",
+    entity="allUsers"
+)
+
+for file in ["404.html", "index.html"]:
+    filepath = os.path.join("wwwroot", file)
+    gcp.storage.BucketObject(
+        file,
+        bucket=bucket.name,
+        name=file,
+        source=pulumi.FileAsset(filepath),
+        opts=pulumi.ResourceOptions(depends_on=[acl])
+    )
+
+static_site_url = pulumi.Output.concat(
+    "https://storage.googleapis.com/", bucket.name, "/index.html")
+
+pulumi.export("static_url", static_site_url)
+```
+
+## Next Steps
```

Comment:
If this infra isn't going to be used in the next lab as a baseline, run a `pulumi destroy`. People won't know to do that and will have leftover infra that will get charged over time.

Reply:
👍 Will be resolved in next commit.